### PR TITLE
Detailed Authorization layers for `ModelView`. Now more customizable.

### PR DESCRIPTION
**Pull request by hasansezertasan**
Related to #443, still in progress, open to reviews and recommendations.

My motivation:
- `get_edit_card_title`, `get_create_card_title` and `get_detail_card_title` methods provides "Python Level" interface configuration, so instead of digging arround Jinja2 templates, developers can easily customize the titles inside the Python Environment.
- `create_button_visible` method enables to disable the "create" button while keeping the create routes and logic.
- `get_one_query` method supports the logic of `get_list_query`, which filters the list page.

### CODE REVIEW

**Comment by hasansezertasan** [comment]
@jowilf, can you check this one and related issue?

Here is the detailed example:

```python
import datetime
from typing import Any, Dict

from sqlalchemy import ForeignKey, create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import RedirectResponse
from starlette.status import HTTP_403_FORBIDDEN
from starlette_admin._types import RequestAction
from starlette.exceptions import HTTPException
from starlette_admin import DateTimeField, HasMany, HasOne, IntegerField, StringField
from starlette_admin._types import RowActionsDisplayType
from starlette_admin.actions import row_action
from starlette_admin.contrib.sqla import Admin, ModelView

engine = create_engine("sqlite:///db.sqlite3", connect_args={"check_same_thread": False})


class Base(DeclarativeBase):
    pass


class Mixin:
    id: Mapped[int] = mapped_column(
        primary_key=True,
        autoincrement=True,
        index=True,
        unique=True,
    )
    date_created: Mapped[datetime.datetime] = mapped_column(
        default=datetime.datetime.utcnow,
        index=True,
    )
    date_updated: Mapped[datetime.datetime] = mapped_column(
        default=datetime.datetime.utcnow,
        onupdate=datetime.datetime.utcnow,
        index=True,
    )


class Shipment(Base, Mixin):
    __tablename__ = "shipment"
    date_to_ship: Mapped[datetime.datetime]
    vehicles: Mapped[list["Vehicle"]] = relationship("Vehicle", back_populates="shipment")


class Vehicle(Base, Mixin):
    __tablename__ = "vehicle"
    make: Mapped[str]
    model: Mapped[str]
    year: Mapped[int]
    shipment_id: Mapped[int] = mapped_column(ForeignKey("shipment.id"))
    shipment: Mapped["Shipment"] = relationship("Shipment", back_populates="vehicles")


class ShipmentView(ModelView):
    row_actions_display_type: RowActionsDisplayType = RowActionsDisplayType.DROPDOWN
    fields = [
        IntegerField(
            name="id",
            label="ID",
            help_text="ID of the record.",
            read_only=True,
        ),
        DateTimeField(
            name="date_created",
            label="Date Created",
            help_text="Date the record was created.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_updated",
            label="Date Updated",
            help_text="Date the record was last updated.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_to_ship",
            label="Date to Ship",
            help_text="Date the shipment should be shipped.",
        ),
        HasMany(
            name="vehicles",
            label="Vehicles",
            help_text="Vehicles in the shipment.",
            identity="vehicle",
            read_only=True,
            exclude_from_create=True,
            exclude_from_edit=True,
        ),
    ]

    @row_action(
        name="add-vehicle",
        text="Add Vehicle by PK",
        icon_class="fa fa-car",
        custom_response=True,
    )
    async def add_vehicle(self, request: Request, pk: Any) -> str:
        url = request.url_for(
            f"{request.app.state.ROUTE_NAME}:create",
            identity="vehicle",
        ).include_query_params(shipment_id=pk)
        return RedirectResponse(url)


class VehicleView(ModelView):
    row_actions_display_type: RowActionsDisplayType = RowActionsDisplayType.DROPDOWN
    add_to_menu: bool = False  # Exclude from the navigation menu no matter what.
    fields = [
        IntegerField(
            name="id",
            label="ID",
            help_text="ID of the record.",
            read_only=True,
        ),
        DateTimeField(
            name="date_created",
            label="Date Created",
            help_text="Date the record was created.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_updated",
            label="Date Updated",
            help_text="Date the record was last updated.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        StringField(
            name="make",
            label="Make",
            help_text="Make of the vehicle.",
            required=True,
        ),
        StringField(
            name="model",
            label="Model",
            help_text="Model of the vehicle.",
            required=True,
        ),
        IntegerField(
            name="year",
            label="Year",
            help_text="Year of the vehicle.",
            required=True,
        ),
        HasOne(
            name="shipment",
            label="Shipment",
            help_text="Shipment the vehicle is in.",
            identity="shipment",
            read_only=True,
            exclude_from_create=True,
            exclude_from_edit=True,
        ),
    ]

    def is_accessible(self, request: Request) -> bool:
        # FIXED: Workaround for hiding the `vehicle` tab from the menu but it's still visible if you navigate to `admin/vehicle/detail/1`, or `admin/vehicle/edit/1`, or `vehicle/create?shipment_id=1`
        # FIXED: So we need to two more logic, displaying the tab (add_to_menu) and serving the list view (can_view_list).
        action: RequestAction = getattr(request.state, "action", None)
        if action in [RequestAction.CREATE, RequestAction.DETAIL, RequestAction.EDIT]:
            return True
        return False

    def can_create(self, request: Request) -> bool:
        # Workaround for removing the create button from the list view.
        # FIXED: It doesn't work if you navigate to `admin/vehicle/list?shipment_id=1`
        # FIXED: So we need to separate the logic of serving the "create" page and displaying the "create" button, which they are united in the `can_create` method.
        shipment_id = request.query_params.get("shipment_id", None)
        print(request.path_params.get("identity", None))
        if not shipment_id:
            return False
        # TODO: PS: We should check if there is a shipment with the given ID.
        return True
    
    def can_edit(self, request: Request) -> bool:
        # FIXED: We don't need to do anything special but the `Save and add another` button is still displayed but gives a 403 error, which is misleading the users.
        # FIXED: Another problem is with the `Cancel` button, it redirects to the list view of the current model, which is not what we want.
        return super().can_edit(request)

    async def before_create(self, request: Request, data: Dict[str, Any], obj: Any) -> None:
        shipment_id = request.query_params.get("shipment_id", None)
        if not shipment_id:
            raise HTTPException(
                status_code=HTTP_403_FORBIDDEN,
                detail="Parent ID is required.",
            )
        # TODO: PS: We should check if there is a shipment with the given ID.
        data["shipment_id"] = shipment_id
        obj.shipment_id = shipment_id
        return await super().before_create(request, data, obj)

    def can_save_and_add_another(self, request: Request) -> bool:
        return False

    def create_cancel_redirect_url(self, request: Request) -> str:
        shpment_id = request.query_params.get("shipment_id", None)
        if not shpment_id:
            return str(request.url_for(f"{request.app.state.ROUTE_NAME}:list", identity="shipment"))
        return str(request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity="shipment", pk=shpment_id))

    def edit_cancel_redirect_url(self, request: Request) -> str:
        shpment_id = request.query_params.get("shipment_id", None)
        if not shpment_id:
            return str(request.url_for(f"{request.app.state.ROUTE_NAME}:list", identity="shipment"))
        return str(request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity="shipment", pk=shpment_id))

    def create_redirect_url(self, request: Request, obj: Any) -> str:
        return str(request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity="shipment", pk=obj.shipment_id))

    def edit_redirect_url(self, request: Request, obj: Any) -> str:
        return str(request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity="shipment", pk=obj.shipment_id))


Base.metadata.create_all(engine)

app = Starlette()
admin = Admin(engine, title="Authorization Hell")
admin.add_view(
    ShipmentView(
        model=Shipment,
    ),
)
admin.add_view(
    VehicleView(
        model=Vehicle,
    ),
)
admin.mount_to(app)

```

**Comment by hasansezertasan** [comment]
I updated the code example in my last comment.

One question remains, how to navigate the browser to the "shipment" list/detail page with a "success" or "failure" message?

File: starlette_admin/views.py [hunk 1]
```diff
@@ -667,18 +670,62 @@ def can_view_details(self, request: Request) -> bool:
...
+        """Permission for saving and continue editing Items. Returns `can_edit` by default"""
+        return self.can_edit(request)
+
     def can_delete(self, request: Request) -> bool:
         """Permission for deleting Items. Return True by default"""
         return True
 
+    def get_detail_card_title(self, request: Request, obj: Any) -> str:
+        """Return the title of the detail card"""
+        return gettext("Details of #%(pk)s") % {"pk": obj[self.pk_attr]}
```
Comment (hasansezertasan) [inline starlette_admin/views.py:699]:
```suggestion
        return gettext("Details of #%(pk)") % {"pk": obj[self.pk_attr]}
```

File: starlette_admin/views.py [hunk 2]
```diff
@@ -667,18 +670,62 @@ def can_view_details(self, request: Request) -> bool:
...
+        """Permission for saving and continue editing Items. Returns `can_edit` by default"""
+        return self.can_edit(request)
+
     def can_delete(self, request: Request) -> bool:
         """Permission for deleting Items. Return True by default"""
         return True
 
+    def get_detail_card_title(self, request: Request, obj: Any) -> str:
+        """Return the title of the detail card"""
+        return gettext("Details of #%(pk)") % {"pk": obj[self.pk_attr]}
```
Comment (hasansezertasan) [inline starlette_admin/views.py:699]: I changed the text for this area, I can replace this line as below, it'll be like old days...

```suggestion
        return gettext("#%(pk)") % {"pk": obj[self.pk_attr]}
```

**Comment by hasansezertasan** [comment]
I updated the code example in my last comment, please check it. It's about this question... How to navigate the client to the "shipment" list/detail page with a "success" or "failure" message on vehicle deletes? The example contains a "row_action" that overriders delete action.

WDYT @jowilf?

---

Another question is this: should we leave "add_to_menu" as a parameter or a function?

---

Yet another question: These lines are identicle with old `model.get_create_cancel_redirect_url(request)` response.

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/edit.html#L10

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/create.html#L10

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/detail.html#L11

What should we do here, something like "is_breadcrumb_available" would be OK?

**Comment by hasansezertasan** [comment]
Code example:

```python
import datetime
from typing import Any, Dict, List

from sqlalchemy import ForeignKey, create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import RedirectResponse
from starlette.status import HTTP_403_FORBIDDEN
from starlette_admin._types import RequestAction
from starlette.exceptions import HTTPException
from starlette_admin import DateTimeField, HasMany, HasOne, IntegerField, StringField
from starlette_admin._types import RowActionsDisplayType
from starlette_admin.actions import row_action
from starlette_admin.contrib.sqla import Admin, ModelView

engine = create_engine("sqlite:///db.sqlite3", connect_args={"check_same_thread": False})


class Base(DeclarativeBase):
    pass


class Mixin:
    id: Mapped[int] = mapped_column(
        primary_key=True,
        autoincrement=True,
        index=True,
        unique=True,
    )
    date_created: Mapped[datetime.datetime] = mapped_column(
        default=datetime.datetime.utcnow,
        index=True,
    )
    date_updated: Mapped[datetime.datetime] = mapped_column(
        default=datetime.datetime.utcnow,
        onupdate=datetime.datetime.utcnow,
        index=True,
    )


class Shipment(Base, Mixin):
    __tablename__ = "shipment"
    date_to_ship: Mapped[datetime.datetime]
    vehicles: Mapped[list["Vehicle"]] = relationship("Vehicle", back_populates="shipment")


class Vehicle(Base, Mixin):
    __tablename__ = "vehicle"
    make: Mapped[str]
    model: Mapped[str]
    year: Mapped[int]
    shipment_id: Mapped[int] = mapped_column(ForeignKey("shipment.id"))
    shipment: Mapped["Shipment"] = relationship("Shipment", back_populates="vehicles")


class ShipmentView(ModelView):
    row_actions_display_type: RowActionsDisplayType = RowActionsDisplayType.DROPDOWN
    fields = [
        IntegerField(
            name="id",
            label="ID",
            help_text="ID of the record.",
            read_only=True,
        ),
        DateTimeField(
            name="date_created",
            label="Date Created",
            help_text="Date the record was created.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_updated",
            label="Date Updated",
            help_text="Date the record was last updated.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_to_ship",
            label="Date to Ship",
            help_text="Date the shipment should be shipped.",
        ),
        HasMany(
            name="vehicles",
            label="Vehicles",
            help_text="Vehicles in the shipment.",
            identity="vehicle",
            read_only=True,
            exclude_from_create=True,
            exclude_from_edit=True,
        ),
    ]

    @row_action(
        name="add-vehicle",
        text="Add Vehicle by PK",
        icon_class="fa fa-car",
        custom_response=True,
    )
    async def add_vehicle(self, request: Request, pk: Any) -> str:
        url = request.url_for(
            f"{request.app.state.ROUTE_NAME}:create",
            identity="vehicle",
        ).include_query_params(shipment_id=pk)
        return RedirectResponse(url)


class VehicleView(ModelView):
    row_actions_display_type: RowActionsDisplayType = RowActionsDisplayType.DROPDOWN
    add_to_menu: bool = False  # Exclude from the navigation menu no matter what.
    parent_column_key: str = "shipment_id"
    parent_model: Any = Shipment
    parent_name: str = "Shipment"
    parent_identity = "shipment"
    
    fields = [
        IntegerField(
            name="id",
            label="ID",
            help_text="ID of the record.",
            read_only=True,
        ),
        DateTimeField(
            name="date_created",
            label="Date Created",
            help_text="Date the record was created.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        DateTimeField(
            name="date_updated",
            label="Date Updated",
            help_text="Date the record was last updated.",
            exclude_from_create=True,
            exclude_from_edit=True,
            read_only=True,
        ),
        StringField(
            name="make",
            label="Make",
            help_text="Make of the vehicle.",
            required=True,
        ),
        StringField(
            name="model",
            label="Model",
            help_text="Model of the vehicle.",
            required=True,
        ),
        IntegerField(
            name="year",
            label="Year",
            help_text="Year of the vehicle.",
            required=True,
        ),
        IntegerField(
            name="shipment_id",
            label="Shipment ID",
            help_text="Shipment ID the vehicle is in.",
            read_only=True,
        ),
        HasOne(
            name="shipment",
            label="Shipment",
            help_text="Shipment the vehicle is in.",
            identity="shipment",
            read_only=True,
            exclude_from_create=True,
            exclude_from_edit=True,
        ),
    ]

    def is_accessible(self, request: Request) -> bool:
        # FIXED: Workaround for hiding the `vehicle` tab from the menu but it's still visible if you navigate to `admin/vehicle/detail/1`, or `admin/vehicle/edit/1`, or `vehicle/create?shipment_id=1`
        # FIXED: So we need to two more logic, displaying the tab (add_to_menu) and serving the list view (can_view_list).
        action: RequestAction = getattr(request.state, "action", None)
        if action not in [RequestAction.CREATE, RequestAction.DETAIL, RequestAction.EDIT, RequestAction.ACTION, RequestAction.ROW_ACTION]:
            return False
        return super().is_accessible(request)

    def can_create(self, request: Request) -> bool:
        # Workaround for removing the create button from the list view.
        # FIXED: It doesn't work if you navigate to `admin/vehicle/list?shipment_id=1`
        # FIXED: So we need to separate the logic of serving the "create" page and displaying the "create" button, which they are united in the `can_create` method.
        parent_column_value = request.query_params.get(self.parent_column_key, None)
        print(request.path_params.get("identity", None))
        if parent_column_value is None:
            return False
        # TODO: PS: We should check if there is a shipment with the given ID.
        return super().can_create(request)
    
    def can_edit(self, request: Request) -> bool:
        # FIXED: We don't need to do anything special but the `Save and add another` button is still displayed but gives a 403 error, which is misleading the users.
        # FIXED: Another problem is with the `Cancel` button, it redirects to the list view of the current model, which is not what we want.
        return super().can_edit(request)

    async def before_create(self, request: Request, data: Dict[str, Any], obj: Any) -> None:
        parent_column_value = request.query_params.get(self.parent_column_key, None)
        if parent_column_value is None:
            raise HTTPException(
                status_code=HTTP_403_FORBIDDEN,
                delatt=f"'{self.parent_column_key}' is missing in the query parameters."
            )
        # TODO: PS: We should check if there is a shipment with the given ID.
        data[self.parent_column_key] = parent_column_value
        setattr(obj, self.parent_column_key, parent_column_value)
        return await super().before_create(request, data, obj)

    def get_create_card_title(self, request: Request) -> str:
        parent_column_value = request.query_params.get(self.parent_column_key, None)
        if parent_column_value is None:
            return "You shouldn't be here!"
        return f"Add {self.name} to {self.parent_name} #{parent_column_value}"

    def get_edit_card_title(self, request: Request, obj: Any) -> str:
        print(obj)
        parent_column_value = obj[self.parent_column_key]
        return f"Edit {self.name} #{obj.id} of {self.parent_name} #{parent_column_value}"

    def can_save_and_add_another(self, request: Request) -> bool:
        return False

    def get_create_cancel_redirect_url(self, request: Request) -> str:
        parent_column_value = request.query_params.get(self.parent_column_key, None)
        if parent_column_value is None:
            return request.url_for(f"{request.app.state.ROUTE_NAME}:list", identity=self.parent_identity)
        return request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity=self.parent_identity, pk=parent_column_value)

    def get_edit_cancel_redirect_url(self, request: Request) -> str:
        parent_column_value = request.query_params.get(self.parent_column_key, None)
        if parent_column_value is None:
            return request.url_for(f"{request.app.state.ROUTE_NAME}:list", identity=self.parent_identity)
        return request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity=self.parent_identity, pk=parent_column_value)

    def get_create_redirect_url(self, request: Request, obj: Any) -> str:
        return request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity=self.parent_identity, pk=obj.shipment_id)

    def get_edit_redirect_url(self, request: Request, obj: Any) -> str:
        return request.url_for(f"{request.app.state.ROUTE_NAME}:detail", identity=self.parent_identity, pk=obj.shipment_id)
    

    @row_action(
        name="delete",
        text="Delete",
        confirmation="Are you sure you want to delete this item?",
        icon_class="fa-solid fa-trash",
        submit_btn_text="Yes, delete",
        submit_btn_class="btn-danger",
        action_btn_class="btn-danger",
        custom_response=True,
    )
    async def row_action_3_delete(self, request: Request, pk: Any) -> str:
        await self.delete(request, [pk])
        return RedirectResponse(str(request.url_for(f"{request.app.state.ROUTE_NAME}:list", identity="shipment")))

Base.metadata.create_all(engine)

app = Starlette()
admin = Admin(engine, title="Authorization Hell")
admin.add_view(
    ShipmentView(
        model=Shipment,
    ),
)
admin.add_view(
    VehicleView(
        model=Vehicle,
    ),
)
admin.mount_to(app)
```

There are still questions:

- How to navigate the client to the "shipment" list/detail page with a "success" or "failure" message on vehicle deletes? The example contains a "row_action" that overriders delete action.
- Should we leave "add_to_menu" as a parameter or turn it into a function?
- These lines are identicle with the default response of `get_create_cancel_redirect_url` method.

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/edit.html#L10

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/create.html#L10

https://github.com/jowilf/starlette-admin/blob/d76582d22e294f756cc3bd6911b6bc138bbd1fb1/starlette_admin/templates/detail.html#L11

What should we do here, something like `is_breadcrumb_available` would be OK or should I implement something like `get_breadcrumb_url` with `get_breadcrumb_label` or both?

- Another one is this:
  - `can_save_and_add_another` removes "Save and add another" button from the UI.
  - `can_save_and_continue_editing` removes "Save and continue editing" button from the UI.
  - `create_button_visible` removes "Create" button for the "list" page.

Their purpose are similar but names are not. What should we do here? Rename `can_save_and_add_nother` as `save_and_add_another_button_visible` or `create_button_visible` as `can_create_button` 😕?

WDYT @jowilf?

**Comment by hasansezertasan** [comment]
@jowilf Waiting to hear from you, if everything looks good, I will also update the documentation.

**Comment by vitasiku** [comment]
Hi @hasansezertasan, did you merge this in? I need this feature.

**Pull request closed without merging**